File: benchmarks/successor/gold/csharp/008-summary-api/OrderFlowEngine.cs
namespace OrderFlow;

public sealed class Customer
{
    public string? Id { get; init; }
    public string? Tier { get; init; }
}

public sealed class Order
{
    public string Id { get; init; } = "";
    public DateTimeOffset CreatedAt { get; init; }
    public string Status { get; init; } = "";
    public int? Priority { get; init; }
    public DateTimeOffset? DueAt { get; init; }
    public Customer? Customer { get; init; }
}

public sealed class Request
{
    public string Operation { get; init; } = "";
    public Order?[]? Orders { get; init; }
    public DateTimeOffset? AsOf { get; init; }
    public string? Id { get; init; }
    public string? ToStatus { get; init; }
}

public sealed record Response(string[] Ids);

public sealed record TransitionResponse(string Id, string Status);

public sealed record SummaryResponse(
    int Pending,
    int Processing,
    int Completed,
    int Cancelled,
    int Overdue);

public static class OrderFlowEngine
{
    public static object Handle(Request request)
    {
        if (IsStatus(request.Operation, "transition"))
        {
            return Transition(request);
        }

        var orders = NormalizeOrders(request.Orders);
        IEnumerable<Order> selected;

        if (IsStatus(request.Operation, "ready"))
        {
            selected = Ready(orders);
        }
        else if (IsStatus(request.Operation, "overdue"))
        {
            var asOf = request.AsOf
                ?? throw new InvalidOperationException("asOf is required for overdue");
            selected = Overdue(orders, asOf);
        }
        else if (IsStatus(request.Operation, "atRisk"))
        {
            var asOf = request.AsOf
                ?? throw new InvalidOperationException("asOf is required for atRisk");
            selected = AtRisk(orders, asOf);
        }
        else if (IsStatus(request.Operation, "vipReady"))
        {
            selected = Ready(orders).Where(IsVip);
        }
        else if (IsStatus(request.Operation, "summary"))
        {
            return Summarize(orders, request.AsOf);
        }
        else
        {
            throw new InvalidOperationException($"Unknown operation: {request.Operation}");
        }

        return new Response(selected.Select(order => order.Id).ToArray());
    }

    private static TransitionResponse Transition(Request request)
    {
        if (string.IsNullOrEmpty(request.Id))
        {
            throw new InvalidOperationException("id is required for transition");
        }

        if (string.IsNullOrEmpty(request.ToStatus))
        {
            throw new InvalidOperationException("toStatus is required for transition");
        }

        var matches = NormalizeOrders(request.Orders)
            .Where(order =>
                string.Equals(order.Id, request.Id, StringComparison.Ordinal))
            .Take(2)
            .ToArray();

        if (matches.Length == 0)
        {
            throw new InvalidOperationException("order not found for transition");
        }

        if (matches.Length > 1)
        {
            throw new InvalidOperationException("duplicate order id for transition");
        }

        var target = CanonicalTarget(matches[0].Status, request.ToStatus);
        return new TransitionResponse(matches[0].Id, target);
    }

    private static string CanonicalTarget(string? source, string target)
    {
        if (IsStatus(source, "pending"))
        {
            if (IsStatus(target, "processing"))
            {
                return "processing";
            }

            if (IsStatus(target, "cancelled"))
            {
                return "cancelled";
            }
        }
        else if (IsStatus(source, "processing"))
        {
            if (IsStatus(target, "completed"))
            {
                return "completed";
            }

            if (IsStatus(target, "cancelled"))
            {
                return "cancelled";
            }
        }

        throw new InvalidOperationException("invalid transition");
    }

    private static SummaryResponse Summarize(
        IEnumerable<Order> orders,
        DateTimeOffset? asOf)
    {
        var pending = 0;
        var processing = 0;
        var completed = 0;
        var cancelled = 0;
        var overdue = 0;

        foreach (var order in orders)
        {
            var isPending = IsStatus(order.Status, "pending");
            var isProcessing = IsStatus(order.Status, "processing");

            if (isPending)
            {
                pending++;
            }
            else if (isProcessing)
            {
                processing++;
            }
            else if (IsStatus(order.Status, "completed"))
            {
                completed++;
            }
            else if (IsStatus(order.Status, "cancelled"))
            {
                cancelled++;
            }

            if (asOf.HasValue
                && (isPending || isProcessing)
                && order.DueAt.HasValue
                && order.DueAt.Value < asOf.Value)
            {
                overdue++;
            }
        }

        return new SummaryResponse(
            pending,
            processing,
            completed,
            cancelled,
            overdue);
    }

    private static Order[] NormalizeOrders(Order?[]? orders) =>
        orders?.OfType<Order>().ToArray() ?? [];

    private static IEnumerable<Order> Ready(IEnumerable<Order> orders) =>
        orders
            .Where(order => IsStatus(order.Status, "pending"))
            .OrderByDescending(order => order.Priority ?? 0)
            .ThenBy(order => order.CreatedAt)
            .ThenBy(order => order.Id, StringComparer.Ordinal);

    private static IEnumerable<Order> Overdue(
        IEnumerable<Order> orders,
        DateTimeOffset asOf) =>
        orders
            .Where(order =>
                IsActive(order.Status)
                && order.DueAt.HasValue
                && order.DueAt.Value < asOf)
            .OrderBy(order => order.DueAt!.Value)
            .ThenByDescending(order => order.Priority ?? 0)
            .ThenBy(order => order.Id, StringComparer.Ordinal);

    private static IEnumerable<Order> AtRisk(
        IEnumerable<Order> orders,
        DateTimeOffset asOf)
    {
        var upperBound = asOf.AddHours(24);

        return orders
            .Where(order =>
                IsActive(order.Status)
                && order.DueAt.HasValue
                && order.DueAt.Value >= asOf
                && order.DueAt.Value < upperBound)
            .OrderBy(order => order.DueAt!.Value)
            .ThenByDescending(order => order.Priority ?? 0)
            .ThenBy(order => order.Id, StringComparer.Ordinal);
    }

    private static bool IsVip(Order order) =>
        order.Customer is not null
        && (IsStatus(order.Customer.Tier, "gold")
            || IsStatus(order.Customer.Tier, "platinum"));

    private static bool IsActive(string? status) =>
        IsStatus(status, "pending") || IsStatus(status, "processing");

    private static bool IsStatus(string? actual, string expected) =>
        string.Equals(actual, expected, StringComparison.OrdinalIgnoreCase);
}

File: benchmarks/successor/gold/fsharp/007-query-engine-refactor/OrderFlowEngine.fs
module OrderFlowEngine

open System

[<CLIMutable>]
type Customer =
    { id: string
      tier: string }

[<CLIMutable>]
type Order =
    { id: string
      createdAt: DateTimeOffset
      status: string
      priority: Nullable<int>
      dueAt: Nullable<DateTimeOffset>
      customer: Customer }

[<CLIMutable>]
type Request =
    { operation: string
      orders: Order array
      asOf: Nullable<DateTimeOffset>
      id: string
      toStatus: string }

[<CLIMutable>]
type Response = { ids: string array }

[<CLIMutable>]
type TransitionResponse =
    { id: string
      status: string }

let private isStatus actual expected =
    String.Equals(actual, expected, StringComparison.OrdinalIgnoreCase)

let private isActive status =
    isStatus status "pending" || isStatus status "processing"

let private priorityOf (order: Order) =
    if order.priority.HasValue then order.priority.Value else 0

let private normalizeOrders (orders: Order array) =
    if Object.ReferenceEquals(orders, null) then
        [||]
    else
        orders
        |> Array.choose (fun order ->
            if isNull (box order) then None else Some order)

let private compareReady (left: Order) (right: Order) =
    let byPriority = compare (priorityOf right) (priorityOf left)

    if byPriority <> 0 then
        byPriority
    else
        let byCreatedAt = DateTimeOffset.Compare(left.createdAt, right.createdAt)

        if byCreatedAt <> 0 then
            byCreatedAt
        else
            StringComparer.Ordinal.Compare(left.id, right.id)

let private compareDuePriorityId (left: Order) (right: Order) =
    let byDueAt = DateTimeOffset.Compare(left.dueAt.Value, right.dueAt.Value)

    if byDueAt <> 0 then
        byDueAt
    else
        let byPriority = compare (priorityOf right) (priorityOf left)

        if byPriority <> 0 then
            byPriority
        else
            StringComparer.Ordinal.Compare(left.id, right.id)

let private ready (orders: Order array) =
    orders
    |> Array.filter (fun order -> isStatus order.status "pending")
    |> Array.sortWith compareReady

let private overdue asOf (orders: Order array) =
    orders
    |> Array.filter (fun order ->
        isActive order.status
        && order.dueAt.HasValue
        && order.dueAt.Value < asOf)
    |> Array.sortWith compareDuePriorityId

let private atRisk (asOf: DateTimeOffset) (orders: Order array) =
    let upperBound = asOf.AddHours(24.0)

    orders
    |> Array.filter (fun order ->
        isActive order.status
        && order.dueAt.HasValue
        && order.dueAt.Value >= asOf
        && order.dueAt.Value < upperBound)
    |> Array.sortWith compareDuePriorityId

let private isVip (order: Order) =
    not (isNull (box order.customer))
    && (isStatus order.customer.tier "gold"
        || isStatus order.customer.tier "platinum")

let private canonicalTarget source target =
    if isStatus source "pending" then
        if isStatus target "processing" then
            "processing"
        elif isStatus target "cancelled" then
            "cancelled"
        else
            raise (InvalidOperationException("invalid transition"))
    elif isStatus source "processing" then
        if isStatus target "completed" then
            "completed"
        elif isStatus target "cancelled" then
            "cancelled"
        else
            raise (InvalidOperationException("invalid transition"))
    else
        raise (InvalidOperationException("invalid transition"))

let private transition (request: Request) : TransitionResponse =
    if String.IsNullOrEmpty(request.id) then
        raise (InvalidOperationException("id is required for transition"))

    if String.IsNullOrEmpty(request.toStatus) then
        raise (InvalidOperationException("toStatus is required for transition"))

    let matches =
        request.orders
        |> normalizeOrders
        |> Array.filter (fun order ->
            String.Equals(order.id, request.id, StringComparison.Ordinal))

    if matches.Length = 0 then
        raise (InvalidOperationException("order not found for transition"))

    if matches.Length > 1 then
        raise (InvalidOperationException("duplicate order id for transition"))

    { id = matches[0].id
      status = canonicalTarget matches[0].status request.toStatus }

let private query (request: Request) : Response =
    let orders = normalizeOrders request.orders

    let selected =
        match request.operation with
        | operation when isStatus operation "ready" ->
            ready orders
        | operation when isStatus operation "overdue" ->
            if not request.asOf.HasValue then
                raise (InvalidOperationException("asOf is required for overdue"))

            overdue request.asOf.Value orders
        | operation when isStatus operation "atRisk" ->
            if not request.asOf.HasValue then
                raise (InvalidOperationException("asOf is required for atRisk"))

            atRisk request.asOf.Value orders
        | operation when isStatus operation "vipReady" ->
            orders
            |> ready
            |> Array.filter isVip
        | operation ->
            raise (InvalidOperationException($"Unknown operation: {operation}"))

    selected
    |> Array.map (fun order -> order.id)
    |> fun ids -> { ids = ids }

let handle (request: Request) : obj | null =
    if isStatus request.operation "transition" then
        box (transition request)
    else
        box (query request)
